### DISCUSSION TITLE
Spacy model loading fails after installing progmatically on Heroku

### DISCUSSION THREAD

**Discussion by Xevion** [explosion/spaCy#10583] [Help: Installation] [answered]
Hi, I'm using the `profanity-filter` package on PyPi which uses `spaCy`. For local development, I've been able to use the `spacy download en` command fine, but since Heroku wipes on every deploy, and because I use `pipenv` which does not have any way to download Spacy models or specify some kind of pre-deploy script to download Spacy, I found [this](https://stackoverflow.com/a/70432019/6912830) post which offered a python script to simply download it pragmatically.

- [Logs](https://gist.github.com/Xevion/9d1cb5a23c3439e7c515e6afd495a6b1)
- [Relevant Source Code](https://github.com/Xevion/runnerspace/blob/3ab3fd547605c65b7831ae5569756e39ad81ba66/app.py#L28-L36)

I implemented this and had it download right before it loaded the rest of my Flask app. Unfortunately, despite the logs showing a completely successful download and install, `profanity-filter` was not able to load the spacy model for some reason.

Are there any issues with how I did this? Can you guys help me?
Thanks for any insights you can provide.

**Comment by polm** [comment, marked as answer]
Our ability to help with third-party packages or debugging application logs is limited, but it looks like you're downloading the model and attempting to use it in the same process, which generally doesn't work. After installation you should restart the Python process. See [here](https://spacy.io/usage/models/#download-python) in the docs, and scroll down for a way to specify models in a `requirements.txt` file (no idea if it'll work with pipenv).

**Reply by Xevion** [reply]
It appears [pipenv DOES support it](https://github.com/pypa/pipenv/issues/3058) although at one point it didn't.
However, I can't figure out how to get spaCy to give me the URL to feet to pipenv.
How do I find that URL?

Thanks.

**Reply by Xevion** [reply]
Nevermind, I found it in the logs - but it still does perk my curiosity into why the docs go into so much detail about downloading directly from URLs but don't offer any specific way of retrieving the URL; it shows in the logs when you request a download and such - but that doesn't seem very smart for a number of obvious reasons...  ?

**Reply by Xevion** [reply]
Still fails, unfortunately, `pipenv` sucessfully installs dependencies with the latest hash on Heroku, I can confirm that - and that Pipfile.lock has `en-core-web-sm` in it, version `2.3.1` just like it wants. I got no clue now.

**Reply by polm** [reply]
Sorry to hear that didn't fix it. As I mentioned before, our ability to support v2 or third-party projects is limited, but if you can narrow down a relevant error message in your logs we may be able to help. Normally I'd suggest you ask the plugin developer, but it seems the code is abandoned and the repo is archived, so I'm not sure much can be done about that.

**Reply by Xevion** [reply]
> Sorry to hear that didn't fix it. ... Normally I'd suggest you ask the plugin developer, but it seems the code is abandoned and the repo is archived, so I'm not sure much can be done about that.

It seems that way. I've given up on it because of that and given that spaCy
is used; this project's performance is probably going to struggle running a
full NLP/text processing model or whatever you wanna call it. I only wanted
to make sure nobody could swear on my resume project; I ended up pivoting
to a different library to get the job done that doesn't need spaCy.

**Reply by polm** [reply]
If all you're trying to do is prevent profanity I'd agree that's overkill. Good luck with your project.
